src/tree/node.rs

```diff
@@ -0,0 +1,384 @@
+use std::cmp::Ordering;
+use std::cmp::Ordering::*;
+use std::mem::{self, replace, swap};
+
+use compare::Compare;
+
+// Nodes keep track of their level in the tree, starting at 1 in the
+// leaves and with a red child sharing the level of the parent.
+#[derive(Clone)]
+pub struct Node<K, V> {
+    key: K,
+    value: V,
+    left: Option<Box<Node<K, V>>>,
+    right: Option<Box<Node<K, V>>>,
+    level: usize,
+}
+
+impl<K, V> Node<K, V> {
+    pub fn key(&self) -> &K { &self.key }
+
+    pub fn value(&self) -> &V { &self.value }
+
+    pub fn value_mut(&mut self) -> &mut V { &mut self.value }
+
+    pub fn key_value_mut(&mut self) -> (&K, &mut V) { (&self.key, &mut self.value) }
+}
+
+impl<K, V> Node<K, V> {
+    #[inline]
+    fn new(key: K, value: V) -> Node<K, V> {
+        Node {key: key, value: value, left: None, right: None, level: 1 }
+    }
+
+    // Remove left horizontal link by rotating right
+    fn skew(&mut self) {
+        if self.left.as_ref().map_or(false, |x| x.level == self.level) {
+            let mut save = self.left.take().unwrap();
+            swap(&mut self.left, &mut save.right); // save.right now None
+            swap(self, &mut save);
+            self.right = Some(save);
+        }
+    }
+
+    // Remove dual horizontal link by rotating left and increasing level of
+    // the parent
+    fn split(&mut self) {
+        if self.right.as_ref().map_or(false,
+          |x| x.right.as_ref().map_or(false, |y| y.level == self.level)) {
+            let mut save = self.right.take().unwrap();
+            swap(&mut self.right, &mut save.left); // save.left now None
+            save.level += 1;
+            swap(self, &mut save);
+            self.left = Some(save);
+        }
+    }
+}
+
+pub fn traverse<K, V, F>(mut node: &Option<Box<Node<K, V>>>, mut f: F)
+    -> &Option<Box<Node<K, V>>> where F: FnMut(&Node<K, V>) -> Ordering {
+
+    while let Some(ref r) = *node {
+        node = match f(r) {
+            Less => &r.left,
+            Greater => &r.right,
+            Equal => break,
+        };
+    }
+
+    node
+}
+
+pub fn traverse_mut<K, V, F>(node: &mut Option<Box<Node<K, V>>>, f: F)
+    -> &mut Option<Box<Node<K, V>>> where F: FnMut(&Node<K, V>) -> Ordering {
+
+    let node = traverse(node, f);
+    unsafe { mem::transmute(node) }
```

Comment:
It should be. The reference we're given is mutable, but it isn't modified
for the duration of the loan.

On Sunday, February 15, 2015, Alexis Beingessner <notifications@github.com>
wrote:

> In src/tree/node.rs
> <https://github.com/Gankro/collect-rs/pull/120#discussion_r24719989>:
>
> > +    while let Some(ref r) = *node {
> > +        node = match f(r) {
> > +            Less => &r.left,
> > +            Greater => &r.right,
> > +            Equal => break,
> > +        };
> > +    }
> > +
> > +    node
> > +}
> > +
> > +pub fn traverse_mut<K, V, F>(node: &mut Option<Box<Node<K, V>>>, f: F)
> > +    -> &mut Option<Box<Node<K, V>>> where F: FnMut(&Node<K, V>) -> Ordering {
> > +
> > +    let node = traverse(node, f);
> > +    unsafe { mem::transmute(node) }
>
> Giving me flashbacks to my first week with Rust :)
>
> Can't remember if this was deemed "sound" or not.
>
> It's definitely... sketchy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Gankro/collect-rs/pull/120/files#r24719989>.
>